File: src/nowifi/monitor.py
"""Monitor mode management for WiFi interfaces.

Handles enabling/disabling monitor mode on macOS and Linux.
Reverts to managed mode on exit (via MonitorGuard context manager).

macOS: Built-in card doesn't support monitor mode. Requires external USB adapter.
Linux: airmon-ng (preferred) or iw dev set type monitor.
"""

from __future__ import annotations

import re
import shutil
import subprocess
import sys
from dataclasses import dataclass


@dataclass
class MonitorInterface:
    """Represents a WiFi interface in monitor mode."""
    name: str           # Monitor mode interface (e.g., wlan0mon)
    original_name: str  # Original managed mode name (e.g., wlan0)
    was_managed: bool   # True if we switched from managed mode


def check_monitor_support(interface: str) -> bool:
    """Check if an interface supports monitor mode without enabling it."""
    if sys.platform == "darwin":
        # macOS built-in WiFi (en0) doesn't support monitor mode
        # External USB adapters might — check if interface exists and isn't en0
        if interface == "en0":
            return False
        # Check if interface exists
        try:
            result = subprocess.run(
                ["ifconfig", interface], capture_output=True, text=True, timeout=5,
            )
            return result.returncode == 0
        except (subprocess.TimeoutExpired, OSError):
            return False

    elif sys.platform == "linux":
        # Check iw for monitor mode support
        try:
            result = subprocess.run(
                ["iw", "phy"], capture_output=True, text=True, timeout=5,
            )
            # Find the phy for our interface
            phy = _get_phy_for_interface(interface)
            if not phy:
                return False
            # Check if "monitor" is in supported interface modes
            in_phy = False
            in_modes = False
            for line in result.stdout.splitlines():
                if phy in line:
                    in_phy = True
                if in_phy and "Supported interface modes:" in line:
                    in_modes = True
                if in_modes and "* monitor" in line:
                    return True
                if in_modes and line.strip() and not line.strip().startswith("*"):
                    in_modes = False
        except (subprocess.TimeoutExpired, OSError, FileNotFoundError):
            pass

        # Fallback: check airmon-ng
        airmon = shutil.which("airmon-ng")
        if airmon:
            try:
                result = subprocess.run(
                    [airmon, "--help"], capture_output=True, text=True, timeout=5,
                )
                return True  # airmon-ng exists, assume it can handle the interface
            except (subprocess.TimeoutExpired, OSError):
                pass

    return False


def find_monitor_interfaces() -> list[str]:
    """Find WiFi interfaces that may support monitor mode."""
    interfaces = []

    if sys.platform == "darwin":
        # On macOS, only external USB WiFi adapters support monitor mode
        # Look for non-standard WiFi interfaces
        try:
            result = subprocess.run(
                ["ifconfig", "-l"], capture_output=True, text=True, timeout=5,
            )
            for iface in result.stdout.strip().split():
                # Skip standard interfaces
                if iface.startswith(("lo", "gif", "stf", "bridge", "utun", "awdl", "llw", "ap")):
                    continue
                if iface == "en0":
                    continue  # Built-in, no monitor mode
                # Check if it's a WiFi interface
                if check_monitor_support(iface):
                    interfaces.append(iface)
        except (subprocess.TimeoutExpired, OSError):
            pass

    elif sys.platform == "linux":
        # Check all wireless interfaces
        try:
            result = subprocess.run(
                ["iw", "dev"], capture_output=True, text=True, timeout=5,
            )
            for match in re.finditer(r"Interface\s+(\S+)", result.stdout):
                iface = match.group(1)
                if check_monitor_support(iface):
                    interfaces.append(iface)
        except (subprocess.TimeoutExpired, OSError, FileNotFoundError):
            # Fallback: check /proc/net/wireless
            try:
                with open("/proc/net/wireless") as f:
                    for line in f:
                        m = re.match(r"\s*(\S+):", line)
                        if m:
                            interfaces.append(m.group(1))
            except (FileNotFoundError, PermissionError):
                pass

    return interfaces


def enable_monitor_mode(interface: str) -> MonitorInterface:
    """Enable monitor mode on a WiFi interface.

    Linux: Uses airmon-ng (preferred) or iw.
    macOS: Limited support — only with external USB adapters.

    Returns MonitorInterface with the monitor-mode interface name.
    Raises RuntimeError if monitor mode cannot be enabled.
    """
    if sys.platform == "linux":
        return _enable_monitor_linux(interface)
    elif sys.platform == "darwin":
        return _enable_monitor_macos(interface)
    else:
        raise RuntimeError(f"Monitor mode not supported on {sys.platform}")


def disable_monitor_mode(mon: MonitorInterface) -> bool:
    """Revert interface from monitor mode to managed mode."""
    if not mon.was_managed:
        return True  # Nothing to revert

    if sys.platform == "linux":
        return _disable_monitor_linux(mon)
    elif sys.platform == "darwin":
        return _disable_monitor_macos(mon)
    return False


def _enable_monitor_linux(interface: str) -> MonitorInterface:
    """Enable monitor mode on Linux."""
    # Try airmon-ng first (handles driver quirks, kills interfering processes)
    airmon = shutil.which("airmon-ng")
    if airmon:
        try:
            # Kill interfering processes
            subprocess.run(
                ["sudo", airmon, "check", "kill"],
                capture_output=True, text=True, timeout=10,
            )
            # Start monitor mode
            result = subprocess.run(
                ["sudo", airmon, "start", interface],
                capture_output=True, text=True, timeout=15,
            )
            # Parse output for new interface name (e.g., wlan0mon)
            output = result.stdout + result.stderr
            m = re.search(r"\(monitor mode.*enabled on (\S+)\)", output)
            if m:
                return MonitorInterface(name=m.group(1), original_name=interface, was_managed=True)
            # Some versions just append "mon"
            mon_name = interface + "mon"
            try:
                check = subprocess.run(
                    ["ifconfig", mon_name], capture_output=True, timeout=3,
                )
                if check.returncode == 0:
                    return MonitorInterface(name=mon_name, original_name=interface, was_managed=True)
            except (subprocess.TimeoutExpired, OSError):
                pass
        except (subprocess.TimeoutExpired, OSError):
            pass

    # Fallback: iw
    try:
        subprocess.run(
            ["sudo", "ip", "link", "set", interface, "down"],
            capture_output=True, timeout=5,
        )
        subprocess.run(
            ["sudo", "iw", "dev", interface, "set", "type", "monitor"],
            check=True, capture_output=True, timeout=5,
        )
        subprocess.run(
            ["sudo", "ip", "link", "set", interface, "up"],
            capture_output=True, timeout=5,
        )
        return MonitorInterface(name=interface, original_name=interface, was_managed=True)
    except (subprocess.CalledProcessError, subprocess.TimeoutExpired, OSError) as e:
        raise RuntimeError(f"Failed to enable monitor mode on {interface}: {e}")


def _enable_monitor_macos(interface: str) -> MonitorInterface:
    """Enable monitor mode on macOS (limited — external adapters only)."""
    if interface == "en0":
        raise RuntimeError(
            "macOS built-in WiFi (en0) does not support monitor mode. "
            "Use an external USB WiFi adapter (recommended: Alfa AWUS036ACH with RTL8812AU)."
        )
    # For external adapters, try setting monitor mode via ifconfig
    try:
        subprocess.run(
            ["sudo", "ifconfig", interface, "monitor"],
            check=True, capture_output=True, timeout=10,
        )
        return MonitorInterface(name=interface, original_name=interface, was_managed=True)
    except (subprocess.CalledProcessError, subprocess.TimeoutExpired, OSError) as e:
        raise RuntimeError(f"Failed to enable monitor mode on {interface}: {e}")


def _disable_monitor_linux(mon: MonitorInterface) -> bool:
    """Disable monitor mode on Linux."""
    airmon = shutil.which("airmon-ng")
    if airmon:
        try:
            subprocess.run(
                ["sudo", airmon, "stop", mon.name],
                capture_output=True, text=True, timeout=15,
            )
            # Restart NetworkManager if it was killed
            nm = shutil.which("systemctl")
            if nm:
                subprocess.run(
                    ["sudo", "systemctl", "restart", "NetworkManager"],
                    capture_output=True, timeout=10,
                )
            return True
        except (subprocess.TimeoutExpired, OSError):
            pass

    # Fallback: iw
    try:
        subprocess.run(
            ["sudo", "ip", "link", "set", mon.name, "down"],
            capture_output=True, timeout=5,
        )
        subprocess.run(
            ["sudo", "iw", "dev", mon.name, "set", "type", "managed"],
            capture_output=True, timeout=5,
        )
        subprocess.run(
            ["sudo", "ip", "link", "set", mon.original_name, "up"],
            capture_output=True, timeout=5,
        )
        return True
    except (subprocess.TimeoutExpired, OSError):
        return False


def _disable_monitor_macos(mon: MonitorInterface) -> bool:
    """Disable monitor mode on macOS."""
    try:
        subprocess.run(
            ["sudo", "ifconfig", mon.name, "-monitor"],
            capture_output=True, timeout=10,
        )
        return True
    except (subprocess.TimeoutExpired, OSError):
        return False


def _get_phy_for_interface(interface: str) -> str:
    """Get the phy name for a wireless interface on Linux."""
    try:
        result = subprocess.run(
            ["iw", "dev", interface, "info"],
            capture_output=True, text=True, timeout=5,
        )
        m = re.search(r"wiphy\s+(\d+)", result.stdout)
        if m:
            return f"phy#{m.group(1)}"
    except (subprocess.TimeoutExpired, OSError, FileNotFoundError):
        pass
    return ""


class MonitorGuard:
    """Context manager that enables monitor mode and reverts on exit.

    Usage:
        with MonitorGuard("wlan0") as mon:
            # mon.name is the monitor interface (e.g., wlan0mon)
            run_capture(mon.name)
        # Automatically reverted to managed mode
    """

    def __init__(self, interface: str):
        self.interface = interface
        self.monitor: MonitorInterface | None = None

    def __enter__(self) -> MonitorInterface:
        self.monitor = enable_monitor_mode(self.interface)
        return self.monitor

    def __exit__(self, *args: object) -> None:
        if self.monitor and self.monitor.was_managed:
            disable_monitor_mode(self.monitor)
            self.monitor = None

File: go/internal/monitor/monitor.go
// Copyright (C) 2026 Mikko Parkkola. All rights reserved.
// Licensed under AGPL-3.0. See LICENSE file.

// Package monitor manages WiFi monitor mode on macOS and Linux.
//
// Handles enabling/disabling monitor mode and reverts to managed mode
// on exit via the Guard type.
//
// macOS: Built-in card does not support monitor mode. Requires an external
// USB WiFi adapter (e.g., Alfa AWUS036ACH with RTL8812AU driver).
//
// Linux: airmon-ng (preferred) or iw dev set type monitor.
package monitor

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"runtime"
	"strings"
)

// Interface represents a WiFi interface that has been put into monitor mode.
type Interface struct {
	Name         string // Monitor mode interface name (e.g., wlan0mon)
	OriginalName string // Original managed mode name (e.g., wlan0)
	WasManaged   bool   // True if we switched from managed mode
}

// ---------------------------------------------------------------------------
// Check support
// ---------------------------------------------------------------------------

// CheckSupport checks if an interface supports monitor mode without enabling it.
func CheckSupport(iface string) bool {
	if runtime.GOOS == "darwin" {
		return checkSupportDarwin(iface)
	}
	if runtime.GOOS == "linux" {
		return checkSupportLinux(iface)
	}
	return false
}

// checkSupportDarwin checks monitor mode support on macOS.
// The built-in WiFi (en0) never supports monitor mode; external USB adapters might.
func checkSupportDarwin(iface string) bool {
	if iface == "en0" {
		return false
	}
	// Check if interface exists.
	cmd := exec.Command("ifconfig", iface)
	if err := cmd.Run(); err != nil {
		return false
	}
	return true
}

// checkSupportLinux checks monitor mode support on Linux via iw phy.
func checkSupportLinux(iface string) bool {
	// Get the phy for this interface.
	phy := getPhyForInterface(iface)
	if phy == "" {
		goto fallback
	}

	{
		out, err := exec.Command("iw", "phy").Output()
		if err == nil {
			inPhy := false
			inModes := false
			for _, line := range strings.Split(string(out), "\n") {
				if strings.Contains(line, phy) {
					inPhy = true
				}
				if inPhy && strings.Contains(line, "Supported interface modes:") {
					inModes = true
					continue
				}
				if inModes {
					trimmed := strings.TrimSpace(line)
					if strings.Contains(trimmed, "* monitor") {
						return true
					}
					// End of modes section: non-empty line that doesn't start with *.
					if trimmed != "" && !strings.HasPrefix(trimmed, "*") {
						inModes = false
					}
				}
			}
		}
	}

fallback:
	// Fallback: check if airmon-ng exists.
	if _, err := exec.LookPath("airmon-ng"); err == nil {
		return true // airmon-ng exists, assume it can handle the interface.
	}

	return false
}

// ---------------------------------------------------------------------------
// Find interfaces
// ---------------------------------------------------------------------------

// FindInterfaces finds WiFi interfaces that may support monitor mode.
func FindInterfaces() []string {
	if runtime.GOOS == "darwin" {
		return findInterfacesDarwin()
	}
	if runtime.GOOS == "linux" {
		return findInterfacesLinux()
	}
	return nil
}

// findInterfacesDarwin finds monitor-capable interfaces on macOS.
// Only external USB WiFi adapters support monitor mode.
func findInterfacesDarwin() []string {
	var interfaces []string

	out, err := exec.Command("ifconfig", "-l").Output()
	if err != nil {
		return nil
	}

	skipPrefixes := []string{"lo", "gif", "stf", "bridge", "utun", "awdl", "llw", "ap"}

	for _, iface := range strings.Fields(string(out)) {
		if iface == "en0" {
			continue // Built-in, no monitor mode.
		}

		skip := false
		for _, prefix := range skipPrefixes {
			if strings.HasPrefix(iface, prefix) {
				skip = true
				break
			}
		}
		if skip {
			continue
		}

		if CheckSupport(iface) {
			interfaces = append(interfaces, iface)
		}
	}

	return interfaces
}

// findInterfacesLinux finds monitor-capable wireless interfaces on Linux.
func findInterfacesLinux() []string {
	var interfaces []string

	ifaceRE := regexp.MustCompile(`Interface\s+(\S+)`)

	out, err := exec.Command("iw", "dev").Output()
	if err == nil {
		for _, m := range ifaceRE.FindAllStringSubmatch(string(out), -1) {
			iface := m[1]
			if CheckSupport(iface) {
				interfaces = append(interfaces, iface)
			}
		}
		if len(interfaces) > 0 {
			return interfaces
		}
	}

	// Fallback: check /proc/net/wireless.
	f, err := os.Open("/proc/net/wireless")
	if err != nil {
		return nil
	}
	defer f.Close()

	lineRE := regexp.MustCompile(`^\s*(\S+):`)
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		if m := lineRE.FindStringSubmatch(scanner.Text()); len(m) > 1 {
			interfaces = append(interfaces, m[1])
		}
	}

	return interfaces
}

// ---------------------------------------------------------------------------
// Enable / Disable
// ---------------------------------------------------------------------------

// Enable puts a WiFi interface into monitor mode.
//
// Linux: Uses airmon-ng (preferred) or iw.
// macOS: Limited support -- only external USB adapters.
//
// Returns the Interface with the monitor-mode interface name.
func Enable(iface string) (*Interface, error) {
	switch runtime.GOOS {
	case "linux":
		return enableLinux(iface)
	case "darwin":
		return enableDarwin(iface)
	default:
		return nil, fmt.Errorf("monitor mode not supported on %s", runtime.GOOS)
	}
}

// Disable reverts an interface from monitor mode to managed mode.
func Disable(mon *Interface) error {
	if mon == nil || !mon.WasManaged {
		return nil // Nothing to revert.
	}

	switch runtime.GOOS {
	case "linux":
		if disableLinux(mon) {
			return nil
		}
		return fmt.Errorf("failed to disable monitor mode on %s", mon.Name)
	case "darwin":
		if disableDarwin(mon) {
			return nil
		}
		return fmt.Errorf("failed to disable monitor mode on %s", mon.Name)
	}
	return nil
}

// ---------------------------------------------------------------------------
// Linux implementation
// ---------------------------------------------------------------------------

// enableLinux enables monitor mode on Linux.
func enableLinux(iface string) (*Interface, error) {
	// Try airmon-ng first (handles driver quirks, kills interfering processes).
	if airmon, err := exec.LookPath("airmon-ng"); err == nil {
		// Kill interfering processes.
		_ = exec.Command("sudo", airmon, "check", "kill").Run()

		// Start monitor mode.
		out, _ := exec.Command("sudo", airmon, "start", iface).CombinedOutput()
		output := string(out)

		// Parse output for new interface name (e.g., wlan0mon).
		re := regexp.MustCompile(`\(monitor mode.*enabled on (\S+)\)`)
		if m := re.FindStringSubmatch(output); len(m) > 1 {
			return &Interface{Name: m[1], OriginalName: iface, WasManaged: true}, nil
		}

		// Some versions just append "mon".
		monName := iface + "mon"
		if err := exec.Command("ifconfig", monName).Run(); err == nil {
			return &Interface{Name: monName, OriginalName: iface, WasManaged: true}, nil
		}
	}

	// Fallback: iw.
	if err := exec.Command("sudo", "ip", "link", "set", iface, "down").Run(); err != nil {
		return nil, fmt.Errorf("failed to bring %s down: %w", iface, err)
	}

	if err := exec.Command("sudo", "iw", "dev", iface, "set", "type", "monitor").Run(); err != nil {
		// Try to bring the interface back up before returning error.
		_ = exec.Command("sudo", "ip", "link", "set", iface, "up").Run()
		return nil, fmt.Errorf("failed to enable monitor mode on %s: %w", iface, err)
	}

	if err := exec.Command("sudo", "ip", "link", "set", iface, "up").Run(); err != nil {
		return nil, fmt.Errorf("failed to bring %s up: %w", iface, err)
	}

	return &Interface{Name: iface, OriginalName: iface, WasManaged: true}, nil
}

// enableDarwin enables monitor mode on macOS (limited -- external adapters only).
func enableDarwin(iface string) (*Interface, error) {
	if iface == "en0" {
		return nil, fmt.Errorf(
			"built-in macOS Wi-Fi (en0) does not support monitor mode; " +
				"use an external USB Wi-Fi adapter (recommended: Alfa AWUS036ACH with RTL8812AU)")
	}

	if err := exec.Command("sudo", "ifconfig", iface, "monitor").Run(); err != nil {
		return nil, fmt.Errorf("failed to enable monitor mode on %s: %w", iface, err)
	}

	return &Interface{Name: iface, OriginalName: iface, WasManaged: true}, nil
}

// disableLinux disables monitor mode on Linux.
func disableLinux(mon *Interface) bool {
	// Try airmon-ng first.
	if airmon, err := exec.LookPath("airmon-ng"); err == nil {
		if err := exec.Command("sudo", airmon, "stop", mon.Name).Run(); err == nil {
			// Restart NetworkManager if it was killed.
			if systemctl, err := exec.LookPath("systemctl"); err == nil {
				_ = exec.Command("sudo", systemctl, "restart", "NetworkManager").Run()
			}
			return true
		}
	}

	// Fallback: iw.
	_ = exec.Command("sudo", "ip", "link", "set", mon.Name, "down").Run()
	_ = exec.Command("sudo", "iw", "dev", mon.Name, "set", "type", "managed").Run()
	_ = exec.Command("sudo", "ip", "link", "set", mon.OriginalName, "up").Run()

	return true
}

// disableDarwin disables monitor mode on macOS.
func disableDarwin(mon *Interface) bool {
	err := exec.Command("sudo", "ifconfig", mon.Name, "-monitor").Run()
	return err == nil
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

// getPhyForInterface gets the phy name for a wireless interface on Linux.
func getPhyForInterface(iface string) string {
	out, err := exec.Command("iw", "dev", iface, "info").Output()
	if err != nil {
		return ""
	}

	re := regexp.MustCompile(`wiphy\s+(\d+)`)
	m := re.FindStringSubmatch(string(out))
	if len(m) > 1 {
		return "phy#" + m[1]
	}
	return ""
}

// ---------------------------------------------------------------------------
// Guard -- context manager pattern
// ---------------------------------------------------------------------------

// Guard enables monitor mode on creation and reverts on Close.
//
// Usage:
//
//	g := monitor.NewGuard("wlan0")
//	mon, err := g.Enable()
//	if err != nil { ... }
//	defer g.Close()
//	// use mon.Name for capture
type Guard struct {
	iface   string
	monitor *Interface
	enable  func(string) (*Interface, error)
	disable func(*Interface) error
}

// NewGuard creates a new Guard for the given interface.
func NewGuard(iface string) *Guard {
	return &Guard{iface: iface, enable: Enable, disable: Disable}
}

// Enable puts the interface into monitor mode and returns the Interface.
func (g *Guard) Enable() (*Interface, error) {
	enable := g.enable
	if enable == nil {
		enable = Enable
	}
	mon, err := enable(g.iface)
	if err != nil {
		return nil, err
	}
	g.monitor = mon
	return mon, nil
}

// Close reverts the interface from monitor mode. Safe to call multiple times.
func (g *Guard) Close() error {
	if g.monitor != nil && g.monitor.WasManaged {
		disable := g.disable
		if disable == nil {
			disable = Disable
		}
		err := disable(g.monitor)
		g.monitor = nil
		return err
	}
	return nil
}
